lib/pipeline.js

```diff
@@ -852,6 +853,128 @@ class PipelineModel extends BaseModel {
         return Promise.allSettled([...toCreateOrUpdate, ...toDeactivate]);
     }
 
+    /**
+     * Converts the simplified stages into a more consistent format
+     *
+     * This is because the user can provide the stage information as:
+     *  - {"name": { "jobs": ["job1", "job2", "job3"], "description": "Description" },
+     *     { "name2": { "jobs": ["job4", "job5"] } }
+     *
+     * We will convert it to a more standard format:
+     *  - [{ "name": "name", "jobs": [1, 2, 3], "pipelineId": 123, "description": "value" },
+     *     { "name": "name2", "jobs": [4, 5], "pipelineId": 123 }]
+     * @method convertStages
+     * @param  {Object}     config              config
+     * @param  {Object}     config.jobs         Jobs
+     * @param  {Object}     config.pipeline     Pipeline
+     * @param  {Object}     config.stages       Pipeline stages
+     * @return {Array}                New array with stages after up-converting
+     */
+    async _convertStages({ pipelineId, stages, jobs, pipelineJobs }) {
+        const newStages = [];
+
+        // Convert stages from object to array of objects
+        Object.entries(stages).forEach(([key, value]) => {
+            const newStage = {
+                name: key,
+                pipelineId,
+                ...value
+            };
+            const jobIds = [];
+
+            // Convert the jobNames to jobIds
+            value.jobs.forEach(jobName => {
+                const job = pipelineJobs.find(j => j.name === jobName);
+
+                if (job) {
+                    jobIds.push(job.id);
+                }
+            });
+
+            newStage.jobIds = jobIds;
+
+            delete newStage.jobs; // extra field from yaml parser
+
+            // Check for setup and teardown
+            const setupJobName = getFullStageName({ stageName: key, type: 'setup' });
+            const teardownJobName = getFullStageName({ stageName: key, type: 'teardown' });
+            const setupJobExists = Object.keys(jobs).find(j => j === setupJobName || j === `~${setupJobName}`);
+            const teardownJobExists = Object.keys(jobs).find(j => j === teardownJobName || j === `~${teardownJobName}`);
+
+            newStage.setup = pipelineJobs.find(j => j.name === setupJobExists).id;
+            newStage.teardown = pipelineJobs.find(j => j.name === teardownJobExists).id;
```

Comment:
`setup` and `teardown` jobs must always exist.
Can this be simplified to?
```suggestion
            newStage.setup = pipelineJobs.find(j => j.name === setupJobName).id;
            newStage.teardown = pipelineJobs.find(j => j.name === teardownJobName).id;
```

Reply:
updated